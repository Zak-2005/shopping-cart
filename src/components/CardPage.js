import React, { useState } from "react";
import { useOutlet, useOutletContext } from "react-router-dom";
const CardPage = ({title,price,setItemsInCart,itemsInCart}) =>{
    const [quantity, setQuantity] = useState(0)
    return (
        <div>
        <h1>{title}</h1>
        <h2>{price}</h2>
        <input type="number" placeholder="Quantity:" min={0} onChange={(e)=>setQuantity(parseInt(e.target.value))}/>
        <button onClick={()=>(quantity>=0)?setItemsInCart(itemsInCart+quantity):0}>Add To Cart</button>
        </div>
    )
}

export default CardPage;