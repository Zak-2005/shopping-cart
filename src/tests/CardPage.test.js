import React from "react";
import { render, screen } from "@testing-library/react";
import CardPage from "../components/CardPage";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import { HashRouter } from "react-router-dom";
import { act } from "react-dom/test-utils";
describe("CardPage Component", () => {
  it("Renders CardPage", () => {
    const { container } = render(
      <HashRouter>
        <CardPage
          title="Product Title"
          src="product.png"
          price={9.99}
          itemsInCart={0}
          setItemsInCart={() => {}}
        />
      </HashRouter>
    );
    expect(container).toMatchSnapshot();
  });
  it("Quantity is set when typed, Cart is set to items in cart + quantity when Add to Cart is clicked", async () => {
    const setItemsInCartMock = jest.fn();
    const user = userEvent.setup();

    render(
      <HashRouter>
        <CardPage
          title="Product Title"
          src="product.png"
          price={9.99}
          itemsInCart={0}
          setItemsInCart={setItemsInCartMock}
        />
      </HashRouter>
    );
    const input = screen.getByPlaceholderText("Quantity:");
    const addToCart = screen.getByRole("button", { name: "Add To Cart" });

    await act(async () => {
      await user.type(input, "10", { eventProperties: { keyCode: 49 } });
      await user.click(addToCart);
    });
    expect(setItemsInCartMock).toHaveBeenCalledWith(10);
    expect(setItemsInCartMock).toHaveBeenCalledTimes(1);
  });
  it("Cart is not set when Quantity is not a positive number", async () => {
    const setItemsInCartMock = jest.fn();
    const user = userEvent.setup();

    render(
      <HashRouter>
        <CardPage
          title="Product Title"
          src="product.png"
          price={9.99}
          itemsInCart={0}
          setItemsInCart={setItemsInCartMock}
        />
      </HashRouter>
    );
    const input = screen.getByPlaceholderText("Quantity:");
    const addToCart = screen.getByRole("button", { name: "Add To Cart" });

    await act(async () => {
      await user.type(input, "-10", { eventProperties: { keyCode: 49 } });
      await user.click(addToCart);
    });
    expect(setItemsInCartMock).toHaveBeenCalledTimes(0);
  });
  
});
