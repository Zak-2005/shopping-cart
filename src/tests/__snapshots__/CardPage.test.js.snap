// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPage Component Renders CardPage 1`] = `
<div>
  <div
    class="cardPage"
  >
    <div
      class="cardPageInfo"
    >
      <h1>
        Product Title
      </h1>
      <img
        src="product.png"
      />
      <h2>
        9.99
      </h2>
      <input
        min="0"
        placeholder="Quantity:"
        type="number"
      />
      <button>
        Add To Cart
      </button>
    </div>
  </div>
</div>
`;
